pkg/smokescreen/smokescreen.go

```diff
@@ -386,26 +427,45 @@ func BuildProxy(config *Config) *goproxy.ProxyHttpServer {
 	// Handle CONNECT proxy to TLS & other TCP protocols destination
 	proxy.OnRequest().HandleConnectFunc(func(host string, pctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
 		pctx.UserData = newContext(config, connectProxy, pctx.Req)
-		defer logProxy(config, pctx)
+		pctx.HTTPErrorHandler = HTTPErrorHandler
 
+		// Defer logging the proxy event here because logProxy relies
+		// on state set in handleConnect
+		defer logProxy(config, pctx)
 		defer pctx.Req.Header.Del(traceHeader)
+
 		err := handleConnect(config, pctx)
 		if err != nil {
-			pctx.Resp = rejectResponse(pctx.Req, config, err)
+			pctx.Resp = rejectResponse(pctx, err)
 			return goproxy.RejectConnect, ""
 		}
 		return goproxy.OkConnect, host
 	})
 
+	// Strangely, goproxy can invoke this same function twice for a single HTTP request.
```

Comment:
😬

Reply:
This handler is _very_ weird and hard to reason about.

Reply:
It is! And every time I need to come back to working on something involving this flow I spend too much time reminding myself how the goproxy machinery works. Hopefully this will help the next fortunate soul stumble through this code.